src/dscanner/analysis/incorrect_infinite_range.d

```diff
@@ -13,85 +13,73 @@ import dparse.lexer;
 /**
  * Checks for incorrect infinite range definitions
  */
-final class IncorrectInfiniteRangeCheck : BaseAnalyzer
+extern(C++) class IncorrectInfiniteRangeCheck(AST) : BaseAnalyzerDmd!AST
 {
-	alias visit = BaseAnalyzer.visit;
+	alias visit = BaseAnalyzerDmd!AST.visit;
 
 	mixin AnalyzerInfo!"incorrect_infinite_range_check";
 
 	///
-	this(string fileName, bool skipTests = false)
+	extern(D) this(string fileName)
 	{
-		super(fileName, null, skipTests);
+		super(fileName);
 	}
 
-	override void visit(const StructBody structBody)
+	override void visit(AST.StructDeclaration sd)
 	{
 		inStruct++;
-		structBody.accept(this);
+		super.visit(sd);
 		inStruct--;
 	}
 
-	override void visit(const FunctionDeclaration fd)
+	override void visit(AST.ClassDeclaration cd)
 	{
-		if (inStruct > 0 && fd.name.text == "empty")
-		{
-			line = fd.name.line;
-			column = fd.name.column;
-			fd.accept(this);
-		}
+		inStruct++;
+		super.visit(cd);
+		inStruct--;
 	}
 
-	override void visit(const FunctionBody fb)
+	override void visit(AST.FuncDeclaration fd)
 	{
-		if (fb.specifiedFunctionBody && fb.specifiedFunctionBody.blockStatement !is null)
-			visit(fb.specifiedFunctionBody.blockStatement);
-		else if (fb.shortenedFunctionBody && fb.shortenedFunctionBody.expression !is null)
-			visitReturnExpression(fb.shortenedFunctionBody.expression);
-	}
+		import dmd.astenums : Tbool;
 
-	override void visit(const BlockStatement bs)
-	{
-		if (bs.declarationsAndStatements is null)
-			return;
-		if (bs.declarationsAndStatements.declarationsAndStatements is null)
-			return;
-		if (bs.declarationsAndStatements.declarationsAndStatements.length != 1)
-			return;
-		visit(bs.declarationsAndStatements);
-	}
+		AST.TypeFunction tf = fd.type.isTypeFunction();
+		AST.ReturnStatement rs = fd.fbody ? fd.fbody.isReturnStatement() : null;
+		AST.CompoundStatement cs = fd.fbody ? fd.fbody.isCompoundStatement() : null;
```

Comment:
Having all of these declarations up front is wasteful. For example, if `tf` is null you should just return from the function, but still you are doing a lot of other operations.

I suggest to declare stuff in a lazy manner, just when you need them. I.E:

```d

if (!fd || fd.ident.toString() != "empty")
    return;

auto tf = fd.type.isTypeFunction();
if (!tf || !tf.next || tf.next.ty != bool)
    return;

...
```

Reply:
I agree, I tried to apply your suggestion.